Warnings when compiling community_leiden

_Originally posted by @szhorvat in https://github.com/igraph/igraph/issues/1349#issuecomment-594133566_

There are quite a few warnings from `community_leiden.c` (@vtraag):

```
community_leiden.c:423:17: warning: implicit declaration of function 'igraph_i_vector_binsearch_slice' is invalid in C99 [-Wimplicit-function-declaration]
                igraph_i_vector_binsearch_slice(&cum_trans_diff, r, &chosen_idx, 0, nb_neigh_clusters);
                ^
community_leiden.c:534:62: warning: incompatible pointer types passing 'void (igraph_vector_t *)' (aka 'void (struct igraph_vector_t *)') to parameter of type 'igraph_finally_func_t *' (aka 'void (*)(void *)')
      [-Wincompatible-pointer-types]
    igraph_vector_ptr_set_item_destructor(&refined_clusters, igraph_vector_destroy);
                                                             ^~~~~~~~~~~~~~~~~~~~~
../include/igraph_vector_ptr.h:82:32: note: passing argument to parameter 'func' here
        igraph_finally_func_t *func);
                               ^
community_leiden.c:725:54: warning: incompatible pointer types passing 'void (igraph_vector_t *)' (aka 'void (struct igraph_vector_t *)') to parameter of type 'igraph_finally_func_t *' (aka 'void (*)(void *)')
      [-Wincompatible-pointer-types]
    igraph_vector_ptr_set_item_destructor(&clusters, igraph_vector_destroy);
                                                     ^~~~~~~~~~~~~~~~~~~~~
../include/igraph_vector_ptr.h:82:32: note: passing argument to parameter 'func' here
        igraph_finally_func_t *func);
                               ^
community_leiden.c:739:22: warning: assigning to 'igraph_t *' (aka 'struct igraph_s *') from 'const igraph_t *' (aka 'const struct igraph_s *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    aggregated_graph = graph;
                     ^ ~~~~~
community_leiden.c:740:29: warning: assigning to 'igraph_vector_t *' (aka 'struct igraph_vector_t *') from 'const igraph_vector_t *' (aka 'const struct igraph_vector_t *') discards qualifiers
      [-Wincompatible-pointer-types-discards-qualifiers]
    aggregated_edge_weights = edge_weights;
                            ^ ~~~~~~~~~~~~
community_leiden.c:741:29: warning: assigning to 'igraph_vector_t *' (aka 'struct igraph_vector_t *') from 'const igraph_vector_t *' (aka 'const struct igraph_vector_t *') discards qualifiers
      [-Wincompatible-pointer-types-discards-qualifiers]
    aggregated_node_weights = node_weights;
                            ^ ~~~~~~~~~~~~
community_leiden.c:1044:24: warning: assigning to 'igraph_vector_t *' (aka 'struct igraph_vector_t *') from 'const igraph_vector_t *' (aka 'const struct igraph_vector_t *') discards qualifiers
      [-Wincompatible-pointer-types-discards-qualifiers]
        i_edge_weights = edge_weights;
                       ^ ~~~~~~~~~~~~
community_leiden.c:1058:24: warning: assigning to 'igraph_vector_t *' (aka 'struct igraph_vector_t *') from 'const igraph_vector_t *' (aka 'const struct igraph_vector_t *') discards qualifiers
      [-Wincompatible-pointer-types-discards-qualifiers]
        i_node_weights = node_weights;
                       ^ ~~~~~~~~~~~~
```

 - `igraph_i_vector_binsearch_slice` will be fixed soon by Vincent's PR
 - I just fixed the ones about `igraph_vector_ptr_set_item_destructor`
 - The `const` stuff is fixable, but I didn't want to mess with the code and risk breaking something. Note that the const applies not to the pointer variable, but to the thing it points to. Thus one solution is if we have `const foo *a; foo *b;` to also have a `const foo *c = flag ? a : b` and work with `c`.  

   There's also the brute-force option of casting away the `const`

   Not a huge deal, but `community_leiden.c` is the only file which triggers such warnings.

--- a/src/community_leiden.c
+++ b/src/community_leiden.c
@@ -702,11 +702,10 @@ static int igraph_i_community_leiden_quality(
  * refined partition, using the non-refined partition to create an initial
  * partition for the aggregate network.
  */
-static int igraph_i_community_leiden(
-        const igraph_t *graph,
-        const igraph_vector_t *edge_weights, const igraph_vector_t *node_weights,
-        const igraph_real_t resolution_parameter, const igraph_real_t beta,
-        igraph_vector_t *membership, igraph_integer_t *nb_clusters, igraph_real_t *quality) {
+int igraph_i_community_leiden(const igraph_t *graph,
+                              igraph_vector_t *edge_weights, igraph_vector_t *node_weights,
+                              const igraph_real_t resolution_parameter, const igraph_real_t beta,
+                              igraph_vector_t *membership, igraph_integer_t *nb_clusters, igraph_real_t *quality) {
     igraph_integer_t nb_refined_clusters;
     long int i, c, n = igraph_vcount(graph);
     igraph_t *aggregated_graph, *tmp_graph;
@@ -743,7 +742,7 @@ static int igraph_i_community_leiden(
     IGRAPH_FINALLY(igraph_vector_destroy, &refined_membership);
 
     /* Initialize aggregated graph, weights and membership. */
-    aggregated_graph = graph;
+    aggregated_graph = (igraph_t*)graph;
     aggregated_edge_weights = edge_weights;
     aggregated_node_weights = node_weights;
     aggregated_membership = membership;
@@ -807,6 +806,7 @@ static int igraph_i_community_leiden(
              * the actual clustering */
             if (nb_refined_clusters >= igraph_vcount(aggregated_graph)) {
                 igraph_vector_update(&refined_membership, aggregated_membership);
+                nb_refined_clusters = *nb_clusters;
             }
 
             /* Keep track of aggregate node. */
@@ -1048,7 +1048,7 @@ int igraph_community_leiden(const igraph_t *graph,
         IGRAPH_FINALLY(igraph_vector_destroy, i_edge_weights);
         igraph_vector_fill(i_edge_weights, 1);
     } else {
-        i_edge_weights = edge_weights;
+        i_edge_weights = (igraph_vector_t*)edge_weights;
     }
 
     /* Check edge weights to possibly use default */
@@ -1062,7 +1062,7 @@ int igraph_community_leiden(const igraph_t *graph,
         IGRAPH_FINALLY(igraph_vector_destroy, i_node_weights);
         igraph_vector_fill(i_node_weights, 1);
     } else {
-        i_node_weights = node_weights;
+        i_node_weights = (igraph_vector_t*)node_weights;
     }
 
     /* Perform actual Leiden algorithm */

